### PR TITLE
Adding additional filtering capability

### PR DESCRIPTION
**Pull request by eparubrub**
User workflow changes: 
- Region, Industry, and Experience filter categories have been added
- Filter categories can be selected and de-selected
- Active filter labels are shown (and functional) even if filter categories are de-selected
- Shows count of all users on the right of categories
- Shows count of selected users on the right of categories
- Region is now shown as a "Pill" component to keep consistency with other filters

Code implementation changes:
- Generalizing categories of filters as "Filter"
- Simplifying logic to pull Filter data
- Redoing implementation from `FocusPicker.tsx` to `filters/FilterPicker.tsx`

Tests:
- Manually (visually) tested changes

### CODE REVIEW

File: components/filters/FilterPickerCategory.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  active,
+  onClick,
+}: FilterPickerCategoryProps) {
+  const handleOnClick = (e) => {
+    if (onClick) onClick(e);
+  };
+  return (
+    <div onClick={handleOnClick}>
+      {category}{" "}
+      <svg height="1rem" width="1rem">
```
Comment (tellaho) [inline components/filters/FilterPickerCategory.tsx]: Screen Shot 2022-10-11 at 11 13 52 PM

Small math issue, this sizing of this SVG is slightly off. Right now it's a `1rem` container with a `.4rem` radius circle. Hence the gap between shape and its edge, at `.1rem`. You can fix this pretty easily by using percentage values for the circle's `cx`, `cy`, & `r` values ([docs](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/circle)). Basically setting all three values to `50%`.

This also might be subjective, but any thoughts on bumping this down in size a bit more? I played around and liked what `0.4rem` looked like. It's likely due to how its diameter is a bit more comparable to the weight/width of the text. Any thoughts?

All together, it could look like this:

```
<svg height="0.4rem" width="0.4rem">
  <circle cx="50%" cy="50%" r="50%" />
</svg>
```

which results in:

image

File: components/filters/FilterPickerCategory.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  category,
+  active,
+  onClick,
+}: FilterPickerCategoryProps) {
+  const handleOnClick = (e) => {
+    if (onClick) onClick(e);
+  };
+  return (
+    <div onClick={handleOnClick}>
+      {category}{" "}
```
Comment (tellaho) [inline components/filters/FilterPickerCategory.tsx]: We can get away with avoiding this whitespace hack by adding margin-left to the svg below, a lá
```
svg {
  // ...
  margin-left: 0.5rem;
  // ...
}
```

File: components/filters/FilterPicker.tsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  const [defaultHeight, setDefaultHeight] = useState<number>();
+  const [focusActive, setFocusActive] = useState<boolean>(true);
+  const [industryActive, setIndustryActive] = useState<boolean>();
+  const [regionActive, setRegionActive] = useState<boolean>();
+  const [experienceActive, setExperienceActive] = useState<boolean>();
+  const listRef = useRef<HTMLUListElement>();
+  const listItemsRef = useRef<HTMLLIElement[]>([]);
+  const filterIsSelected = activeFilters.length !== 0;
+
+  useLayoutEffect(() => {
```
Comment (tellaho) [inline components/filters/FilterPicker.tsx:38]: It looks like we removed the limit of two rows per category with the **View All** button. Thinking this `useLayoutEffect` no longer actually does anything now.

**before:**
image

**after:**
image


The behavior was a bit awkward at times but it did limit this situation for smaller screens. Do you think it'd be worth re-adding? Happy to try advising here if you were finding it unwieldy.

File: components/MemberDirectory.tsx [hunk 4]
```diff
@@ -34,14 +48,36 @@ export default function MemberDirectory({ members }: MemberDirectoryProps) {
               <h2 className="member__name">{member.name}</h2>
               <div className="member__location">
                 <h3>{member.location}</h3>
-                <h4>{member.region}</h4>
+                {member.regionFilter[0] ? (
+                  <h4>
+                    <Pill
+                      active={member.regionFilter[0].active}
+                      customWidth="max-content"
+                      customFontSize="1rem"
```
Comment (tellaho) [inline components/MemberDirectory.tsx]: Curious whatcha thinking with this new `customFontSize` prop?

Screen Shot 2022-10-12 at 12 15 59 AM

I worry it creates an inconsistent size between the two pill groups. It looks a bit awkward as they're so close to each other. Especially when their closest adjacent sibling (the location or job title) are otherwise the same.

File: components/filters/FilterPicker.tsx [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    if (!listItemsRef) return;
+    setDefaultHeight(listRef.current.scrollHeight);
+  }, [width]);
+
+  function activateFilter(
```
Comment (tellaho) [inline components/filters/FilterPicker.tsx]: My gut is that these look and feel like tabs, but don't behave like tabs. What do you think about adjusting the behavior to match how tabs more typically behave?

gif

This means:
- **One (and only one) can be selected at a time.** Not more than one or zero.
- **Clicking on one deselects the others.** The only exception is when you click the active tab (which does nothing).

**Review by tellaho — COMMENTED** [review]
What a sprint, brova! Full design to development.

I dropped a few comments that I think impact the UX. Let me know what you think; I'd be happy to chat about any of it. Keep it up. Mahalo for your effort.

File: components/MemberDirectory.tsx [hunk 6]
```diff
@@ -6,6 +6,9 @@ import Pill from "./Pill";
 
 export interface DirectoryMember extends MemberPublic {
   focus: { active?: boolean; id: string; name: string }[];
+  industry: { active?: boolean; id: string; name: string }[];
+  experienceFilter: { active?: boolean; id: string; name: string }[];
```
Comment (ataeoalii) [inline components/MemberDirectory.tsx:10]: One minor thing I just realized is that as this evolves over time, people's experience in tech will change year over year, so we'll probably want to change this to a date instead of a static year range (not super important to this PR tho).

File: components/filters/FilterPicker.tsx [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+export interface PickerFilter extends Filter {
+  active?: boolean;
+}
+
+interface FilterPickerProps {
+  filtersList: PickerFilter[];
+  activeFilters: PickerFilter[];
+  onFilterClick: (id?: string, filterType?: string) => any;
+  onFilterSeclect: (filterSelect?: string, enable?: boolean) => any;
```
Comment (ataeoalii) [inline components/filters/FilterPicker.tsx]: Minor: spelling.
`onFilterSelect`

File: components/filters/FilterPicker.tsx (hunk 5 above)
Comment (ataeoalii) [reply components/filters/FilterPicker.tsx]: If we do allow multiple filter categories to be shown at once, it makes sense to have the most recently selected at the top instead of the bottom (so prepend the newly enabled category instead of append). This allows the user to see what they've selected especially if the screen is smaller, they'll have to scroll to the bottom.

That said, I like the suggestion of treating each category like a tab.

One minor thing, I'm definitely not a designer so take this with a grain of salt, but is there a simple way to indicate which category a selected filter belongs to? For example, if "Engineering" is selected, add a tool tip on hover stating that it belongs to the "Focus" category.

File: lib/api.ts [hunk 8]
```diff
@@ -164,63 +170,105 @@ export async function getMembers(): Promise<MemberPublic[]> {
...
-    .filter((role) => role.fields["Name"])
-    .filter((role) =>
-      limitByMembers ? role.fields["Has Approved Members"] : true
+export async function getFilters(
+  filterType: string,
+  limitByMembers?: boolean,
+  approvedMemberIds?: string[]
+): Promise<Filter[]> {
+  const filters = await getBase({
+    name: filterType == "focus" ? "Focuses" : "Industries",
```
Comment (ataeoalii) [inline lib/api.ts]: nit: should these filterType strings be stored in a variable somewhere?

**Review by ataeoalii — COMMENTED** [review]
Love the added functionality, nice work. I left a few minor comments. Thanks Kamakani.

File: components/filters/FilterPickerCategory.tsx (hunk 1 above)
Comment (eparubrub) [reply components/filters/FilterPickerCategory.tsx]: Yea I was actually looking for a suggestion like that for the `cx`, `cy`, & `r` values. Makes sense to have the x and y values set to 50% in order to center it. 

Also, I totally agree on the sizing and I'll add this in the next commit.

File: components/filters/FilterPickerCategory.tsx (hunk 2 above)
Comment (eparubrub) [reply components/filters/FilterPickerCategory.tsx]: That makes sense, I'll add this in the next commit.

File: components/filters/FilterPicker.tsx (hunk 3 above)
Comment (eparubrub) [reply components/filters/FilterPicker.tsx:38]: @taylorkmho yea this was interesting and something we both talked about during my initial proposals for this. I definitely had that scenario a couple times where I'm like "wait am I seeing all the filtering options? oh right there's a _View All_ option I need to press". But then again I totally see the benefit of being able to hide a lot of it for smaller screens while also showing new users we have this capability.

I tried thinking about this and had some ideas like maybe we have that same type of "limit to two rows" on initial page load, but instead of a _View All_ option, the Focus category would oscilate colors between `theme.color.brand.faded` and `theme.color.brand.base` and the 2 rowed filter options would be slightly faded as well. Then when a user clicked on it, it would open up the rest of the options and act as normal. But this implementation was a bit complex in my head and didn't want to go too deep into the implementation without talking it over.

**TLDR**: I thought about it, had some complex ideas but ultimately thought it would be better to let you advise.

File: components/MemberDirectory.tsx (hunk 4 above)
Comment (eparubrub) [reply components/MemberDirectory.tsx]: @taylorkmho I had a few things in mind with this one: 

1. I was trying to keep it a bit consistent with the sizing of the original Region text, it is slightly bigger than the filter pills on the current implementation

Screen Shot 2022-10-12 at 8 23 51 PM

2. It feels a _tiny_ bit small when it's at the original size in comparison to the **location** right above it. In my eyes, some of the location names aren't as familiar (e.g. Bend or Lehi) with me so it feels like the **region** is more of an aside from the **location** which is visually hard to see when the **region** is much smaller, yet the **region** is the filter that groups most of the locations up. 

Screen Shot 2022-10-12 at 8 26 47 PM

I don't have strong feelings about the sizing though. Maybe changing the sizing or ordering of the location might be better but either way, I'm definitely down to change it so I'll leave the decision up to you.

File: components/filters/FilterPicker.tsx (hunk 5 above)
Comment (eparubrub) [reply components/filters/FilterPicker.tsx]: I agree I like the tab selection, it actually also in theory keeps the user knowing what category of filter they're looking at. I'll add this to the next commit.

Also to your point Andrew, would this tab behavior make it more clear for filter selection? I think your tooltip suggestion would have been most helpful for the multiple filter categories being shown at once but since we're all in agreeance that tab behavior is better I think that solves it? Or maybe I'm not seeing some other benefits you're seeing.

File: components/MemberDirectory.tsx (hunk 6 above)
Comment (eparubrub) [reply components/MemberDirectory.tsx:10]: Hmmm yea this is actually an implementation design from the way we consume new user submissions, notifying @taylorkmho maybe we can discuss this in the future. I'm assuming since this site has been up, a few users already need to shift their years of experience up (assuming they didn't send an update request)

File: components/filters/FilterPicker.tsx (hunk 7 above)
Comment (eparubrub) [reply components/filters/FilterPicker.tsx]: nice catch, I'll add this correction* to the next commit.

File: lib/api.ts (hunk 8 above)
Comment (eparubrub) [reply lib/api.ts]: @ataeoalii are you suggesting something like this? 

Screen Shot 2022-10-12 at 8 53 38 PM

Comment (eparubrub) [reply lib/api.ts]: I'm adding this to the next commit, let me know if that's not what you meant

File: components/filters/FilterPicker.tsx (hunk 5 above)
Comment (eparubrub) [reply components/filters/FilterPicker.tsx]: update is live here: https://hawaiiansintech-53p35qftu-hawaiians.vercel.app/

Comment (ataeoalii) [reply components/filters/FilterPicker.tsx]: Yup looks good, the tool tip idea was more for the combined filter categories. I probably should've changed the order of my comments. Update looks good.

File: components/MemberDirectory.tsx (hunk 6 above)
Comment (ataeoalii) [reply components/MemberDirectory.tsx:10]: I added the issue here: https://github.com/hawaiiansintech/hawaiiansintech/issues/27. It may be simple enough to just keep it as is and have people manually update their profile each year.

File: components/MemberDirectory.tsx [hunk 9]
```diff
@@ -14,12 +17,23 @@ interface MemberDirectoryProps {
...
   const isFiltered =
-    members.filter((mem) => mem.focus?.filter((foc) => foc.active).length > 0)
-      .length > 0;
+    members.filter(
+      (mem) =>
+        mem.focus
+          .concat(mem.industry)
+          .concat(mem.experienceFilter)
+          .concat(mem.regionFilter)
+          ?.filter((foc) => foc.active).length > 0
```
Comment (ataeoalii) [inline components/MemberDirectory.tsx:26]: Curious your thoughts here: for the selected filters, I think we should OR together within the filter category, but AND together separate categories.

For example:

If I select Region: `Hawaii` + `Florida` and Experience: `3-4 years`, I expect to see a list of people in Hawaii or Florida with 3-4 years of experience. I don't expect to see people in Hawaii with 20 years experience.

If I select Region: `Hawaii` and Experience: `3-4 years` + `5-9 years`, I expect to see all people in Hawaii with 3-9 years.

Current behavior shows people that match ANY criteria.

Comment (eparubrub) [reply components/MemberDirectory.tsx:26]: Great point and something I mentioned to @taylorkmho as an idea. I thought there maybe should be a toggle between AND / OR for filter selection like you said, but then I was also thinking about the fact that it already **sorts the results by the highest amount of matched filters**, which I thought might be* enough. 

Then again I'm not too sure though as I could see some scenarios where it appears a bit confusing. The implementation wouldn't be too hard, but I'll leave that decision up to Taylor.

Comment (eparubrub) [reply components/MemberDirectory.tsx:26]: Ah my bad Andrew, just now reading again, you're actually suggesting built-in AND/OR functionality which logically makes sense, I'm just having a hard time conceptualizing how we communicate this filtering logic to users via the UI/UX, or whether or not it's important to

Comment (ataeoalii) [reply components/MemberDirectory.tsx:26]: Actually, that would be even better; I was just saying that the default behavior should OR within a category and AND between categories. An example where this could be helpful, say I'm a recruiter in Oregon looking for someone with 3-5 years of frontend experience seeing a list of 100 people that match any of those search criteria would be misleading when there might only be 5 people that match all of them.

**Review by ataeoalii — APPROVED** [review]
Approving this, feel free to ignore my 1 outstanding comment. I don't think it's critical.

File: components/MemberDirectory.tsx (hunk 9 above)
Comment (tellaho) [reply components/MemberDirectory.tsx:26]: Yeah after playing with it a bit, there's something about filtering by multiple categories that _feels_ like more control or communication may be ideal.

I think I'm with Andrew... I wouldn't let this be what blocks you from merging... since someone should be able to deduce how their filters relate to the results below, especially with how the tags "highlight" if they match.

File: components/MemberDirectory.tsx (hunk 4 above)
Comment (tellaho) [reply components/MemberDirectory.tsx]: Any thoughts on not using a tag for the region? I understand we want to highlight selected criteria or make it obvious, which the Pill pattern comes with (with its "Selected" state). Either way, but I do agree that it does feel a bit "backwards" when it's smaller. I personally read into it as more being due to it being a tag/pill than the size.

File: components/filters/FilterPicker.tsx (hunk 3 above)
Comment (tellaho) [reply components/filters/FilterPicker.tsx:38]: If I were to throw the UX book at you, I'd lean on limiting the number of selections to something reasonable with an obvious control to increase that number. Folks have trouble keeping track of too many options.

But I'd hate for you to have to re-engineer what was built the first time. I imagine enough has changed that it's no longer 1:1 to re-add it.

**Comment by tellaho** [comment]
I personally think you should be ready to merge unless you felt strongly about fixing some of the nits I mentioned above. We can always come back to it; it's feeling pretty dang good as it is.

File: components/filters/FilterPicker.tsx (hunk 3 above)
Comment (eparubrub) [reply components/filters/FilterPicker.tsx:38]: Yup, was reading about Hicks Law in **Laws of UX** on the train the other day - _The time it takes to make a decision increases with the number and complexity of choices available_. Also Miller's Law - _The average person can keep only 7 (+/- 2) items in their working memory_.

I definitely want to come back to this issue in a future PR

File: components/MemberDirectory.tsx (hunk 4 above)
Comment (eparubrub) [reply components/MemberDirectory.tsx]: I don't really have a preference on whether we use a tag or not. I can experiment with a few solutions for this on the next PR.

**Comment by eparubrub** [comment]
Merging with the remaining features being deferred to future PRs:
1. Limit the amount of filter options being presented when a user is first introduced to the site and provide a control to show all of the filter options. ([Refer to this comment](https://github.com/hawaiians/hawaiiansintech/pull/6#discussion_r1002770267))
2. Find a solution to handle Region filtering. Currently the Region is being displayed via Pill but has room for improvement. ([Refer to this comment](https://github.com/hawaiians/hawaiiansintech/pull/6#discussion_r1002769780))
3. Find a solution to multi-filtering controls. Currently filtering across different filter types may feel unintuitive. Some exploration with possibilities should be done with this. ([Refer to this comment](https://github.com/hawaiians/hawaiiansintech/pull/6#discussion_r994241975))
